### PR TITLE
Feat: day 3

### PR DESCRIPTION
**Pull request by jling90**
### Description

Add solutions for day 2 tasks.

To run:

```
cargo run 3 1
cargo run 3 2
```

### CODE REVIEW

File: src/day_3.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+use std::fs::File;
+use std::io::{self};
+
+const N_BITS: usize = 12;
```
Comment (RyanCarrier) [inline src/day_3.rs:5]: This fails the sample data which is 5 bits

-$10,000 from your salary

File: src/day_3.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+      for (index, bit) in bits.char_indices() {
+        if bit == '1' {
+          ones_counts[index] += 1
+        }
+      }
+    }
+  }
+
+  ones_counts.iter().for_each(|n| {
```
Comment (RyanCarrier) [inline src/day_3.rs:29]: you're essentially doing a fold here but you have started your variables way earlier (just do a fold)

File: src/main.rs [hunk 3]
```diff
@@ -44,7 +46,16 @@ fn main() {
...
+            _ => println!("No match for exercise {}, task {}", day_num, task_num),
+        },
+        "3" => match task_num.as_str() {
+            "1" => {
+                if let Ok(lines) = util::read_lines(EX_3_FILE_PATH) {
+                    println!("{}", day_3::task_one(lines));
+                } else {
+                    println!("Failed to read file");
+                }
+            }
```
Comment (RyanCarrier) [inline src/main.rs:58]: ur brain is huge 10/10

this is better than mine

**Review by RyanCarrier — COMMENTED** [review]
🤓🤓🤓🤓

File: src/day_3.rs [hunk 4]
```diff
@@ -8,7 +8,7 @@ const N_BITS: usize = 12;
 // Return as a tuple
 // - String with 1s in each position where 1 is most common.
 // - String with 0s in each position where 1 is most common.
-fn most_popular_bit_strings(lines: Vec<String>) -> (String, String) {
+fn most_popular_bit_strings(lines: &Vec<String>) -> (String, String) {
   // Threshold for determining whether '1s' are 'more common'
   // This is half the total number of lines.
   let midpoint = lines.len() / 2;
```
Comment (RyanCarrier) [inline src/day_3.rs:14]: this is better than ur float thing

File: src/day_3.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  return (gamma, epsilon);
+}
+
+/**
+ * Given a vec of Strings, `strings` and a search String `search`,
+ * Starting with the minimal substring of `search`, find a set of
+ * matching Strings from `strings`. Once a single match exists for
+ * a given substring, return it.
+ */
+fn get_only_substring(strings: &Vec<String>, search: String) -> String {
```
Comment (RyanCarrier) [inline src/day_3.rs:48]: I feel like you have miss-understood this one. Frequency of bits needs to be re-calculated by the remaining bytes.

As such you won't even need the search string, but you will need something to differentiate between oxygen and CO2 as they break ties differently

File: src/day_3.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      return strings_copy[0].to_string();
+    }
+
+    println!(
+      "cursor is {}, strings_copy is {:?}, search is {:?}",
+      cursor, strings_copy, search
+    );
+  }
+
+  // FIXME: Pls no stop why are you doing this to me
```
[the remark this answers is not in the document]
Comment (RyanCarrier) [reply src/day_3.rs:70]: Most common bits need to be re-calculated every filter, currently you are using the most common bits for ALL bytes available initially

File: src/day_3.rs (hunk 1 above)
Comment (jling90) [reply src/day_3.rs:5]: Damn it my salary is negative now

**Review by RyanCarrier — CHANGES_REQUESTED** [review]
Not following the spec, rookie mistake.
You're lucky you're cute.

File: src/day_3.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jling90) [reply src/day_3.rs:70]: oohhhhhhh yep I get what you mean 🧠 💩 - thanks homie

File: src/day_3.rs (hunk 5 above)
Comment (jling90) [reply src/day_3.rs]: https://github.com/jling90/advent-of-code-2021/pull/4/commits/286d47aa2d73b5f242ebe4b088b112d89dbd2bf2 we did it fam

**Pull request merged**